hiclass/Explainer.py

```diff
@@ -0,0 +1,283 @@
+
+        if (
+            isinstance(self.hierarchical_model, LocalClassifierPerParentNode)
+            or isinstance(self.hierarchical_model, LocalClassifierPerLevel)
+            or isinstance(self.hierarchical_model, LocalClassifierPerNode)
+        ):
+            return self._explain_with_xr(X)
+        else:
+            raise ValueError(f"Invalid model: {self.hierarchical_model}.")
+
+    def _explain_with_xr(self, X):
+        """
+        Generate SHAP values for each node using the SHAP package.
+
+        Parameters
+        ----------
+        X : array-like
+            Sample data for which to generate SHAP values.
+
+        Returns
+        -------
+        explanation : xarray.Dataset
+            An xarray Dataset consisting of SHAP values for each sample.
+        """
+        explanations = Parallel(n_jobs=self.n_jobs, backend="threading")(
+            delayed(self._calculate_shap_values)(sample.reshape(1, -1)) for sample in X
+        )
+
+        dataset = xr.concat(explanations, dim="sample")
+        return dataset
+
+    def _get_traversed_nodes(self, samples):
+        """
+        Return a list of all traversed nodes as per the provided HiClass model.
+
+        Parameters
+        ----------
+        samples : array-like
+            Sample data for which to generate traversed nodes.
+
+        Returns
+        -------
+        traversals : list
+            A list of all traversed nodes.
+        """
+        # Helper function to return traversed nodes
+        if isinstance(self.hierarchical_model, LocalClassifierPerParentNode):
+            # Initialize array with shape (#samples, #levels)
+            traversals = np.empty(
+                (samples.shape[0], self.hierarchical_model.max_levels_),
+                dtype=self.hierarchical_model.dtype_,
+            )
+
+            # Initialize first element as root node
+            traversals[:, 0] = self.hierarchical_model.root_
+
+            # For subsequent nodes, calculate mask and find predictions
+            for level in range(1, traversals.shape[1]):
+                predecessors = set(traversals[:, level - 1])
+                predecessors.discard("")
+                for predecessor in predecessors:
+                    mask = np.isin(traversals[:, level - 1], predecessor)
+                    predecessor_x = samples[mask]
+                    if predecessor_x.shape[0] > 0:
+                        successors = list(
+                            self.hierarchical_model.hierarchy_.successors(predecessor)
+                        )
+                        if len(successors) > 0:
+                            classifier = self.hierarchical_model.hierarchy_.nodes[
+                                predecessor
+                            ]["classifier"]
+                            traversals[mask, level] = classifier.predict(
+                                predecessor_x
+                            ).flatten()
+            return traversals
```

Comment:
If there are different implementations of this method for each local classifier, I would recommend to either use inheritance or call another method with the different implementations. Otherwise I am afraid this method will get way too large in the next pull requests.

Reply:
Update: Now instead of `_get_traversed_nodes` I will write three separate method calls  `_get_traversed_nodes_lcppn`, `_get_traversed_nodes_lcpn`, and `_get_traversed_nodes_lcpl`.

Each representing the individual implementation for the corresponding hiclass model. The `_calculate_shap_values` method now calls the appropriate method by a `if-elif` construct which checks the hiclass model using `isinstance() util.

Resolved by [ada4e88](https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/ada4e88880a486099487a81cde0eb9b3ebec59ca).